node/core/pvf/src/execute/queue.rs

```diff
@@ -172,47 +234,80 @@ async fn purge_dead(metrics: &Metrics, workers: &mut Workers) {
 	}
 }
 
+fn try_assign_next_job(queue: &mut Queue) {
+	if let Some(ji) = queue.next_job_index(None) {
+		if let Some(available) =
+			queue.workers.find_available(queue.queue[ji].executor_params.hash())
+		{
+			let job = queue.take_job(ji).expect("Job is just checked to be in queue; qed");
+			assign(queue, available, job);
+			return
+		} else {
+			if let Some(idle) = queue.workers.find_idle() {
+				// No available workers of required type but there are some idle ones of other types,
+				// have to kill one and re-spawn with the correct type
+				if queue.workers.running.remove(idle).is_some() {
+					queue.metrics.execute_worker().on_retired();
+				}
+			}
+		}
+
+		if queue.workers.can_afford_one_more() {
+			let job = queue.take_job(ji).expect("Job is just checked to be in queue; qed");
+			spawn_extra_worker(queue, job);
+		}
+	}
+}
+
 fn handle_to_queue(queue: &mut Queue, to_queue: ToQueue) {
-	let ToQueue::Enqueue { artifact, execution_timeout, params, result_tx } = to_queue;
+	let ToQueue::Enqueue { artifact, execution_timeout, params, executor_params, result_tx } =
+		to_queue;
 	gum::debug!(
 		target: LOG_TARGET,
 		validation_code_hash = ?artifact.id.code_hash,
 		"enqueueing an artifact for execution",
 	);
 	queue.metrics.execute_enqueued();
-	let job = ExecuteJob { artifact, execution_timeout, params, result_tx };
-
-	if let Some(available) = queue.workers.find_available() {
-		assign(queue, available, job);
-	} else {
-		if queue.workers.can_afford_one_more() {
-			spawn_extra_worker(queue);
-		}
-		queue.queue.push_back(job);
-	}
+	let job = ExecuteJob {
+		artifact,
+		execution_timeout,
+		params,
+		executor_params,
+		result_tx,
+		waiting_since: Instant::now(),
+	};
+	queue.queue.push_back(job);
+	try_assign_next_job(queue);
 }
 
 async fn handle_mux(queue: &mut Queue, event: QueueEvent) {
 	match event {
-		QueueEvent::Spawn(idle, handle) => {
-			handle_worker_spawned(queue, idle, handle);
+		QueueEvent::Spawn(idle, handle, job) => {
+			handle_worker_spawned(queue, idle, handle, job);
 		},
 		QueueEvent::StartWork(worker, outcome, artifact_id, result_tx) => {
 			handle_job_finish(queue, worker, outcome, artifact_id, result_tx);
 		},
 	}
 }
 
-fn handle_worker_spawned(queue: &mut Queue, idle: IdleWorker, handle: WorkerHandle) {
+fn handle_worker_spawned(
+	queue: &mut Queue,
+	idle: IdleWorker,
+	handle: WorkerHandle,
+	job: ExecuteJob,
+) {
 	queue.metrics.execute_worker().on_spawned();
 	queue.workers.spawn_inflight -= 1;
-	let worker = queue.workers.running.insert(WorkerData { idle: Some(idle), handle });
+	let worker = queue.workers.running.insert(WorkerData {
+		idle: Some(idle),
+		handle,
+		executor_params_hash: job.executor_params.hash(),
+	});
 
 	gum::debug!(target: LOG_TARGET, ?worker, "execute worker spawned");
 
-	if let Some(job) = queue.queue.pop_front() {
-		assign(queue, worker, job);
-	}
+	assign(queue, worker, job);
```

Comment:
uh I see, this is the reason `job` is passed around. But why this change? Why not look into the queue again similar to how it was? I guess this could give slightly better performance, because we then can already look into the next task while the worker is spawning. If that's the case, might be worth documenting, to avoid confusion.

Reply:
Also I am not sure it is worth it. If there are at least two jobs with the same required execution environment, then this optimization does already nothing.

Reply:
The overall idea of the new implementation is like the following.

Old design: we have workers, assign jobs to them as soon as a worker is free, and spawn a worker if it's dead.

New design: no worker is spawned without a job pre-assigned to it. No jobs, no workers. An idle worker which finishes its job may still be used to execute a job with a compatible execution environment type.

That was supposed to eliminate any race condition. So, for example, we have a job we want to execute, and we want to kill an existing worker and spawn a new one to execute it. Still, while a new worker is spawned, a job already existing in the queue becomes stale and should be executed immediately, and we have to kill a newly created worker and spawn a new one to execute it, and so on. So I wanted workers to have a hard-assigned job beforehand, without possibly falling back.

It seems like I should evaluate that behavior from scratch to be sure it's sound.

Reply:
So you are accounting for the fact that the queue state might change in the meantime and once the worker comes ready, we would immediately decide to kill it again, without having executed even a single job on it. That makes sense (although it looks like the code doing that is never triggered). In any case, please document in code.